db/background_mgr.go

```diff
@@ -52,53 +85,181 @@ type BackgroundManagerStatus struct {
 // BackgroundManagerProcessI is an interface satisfied by any of the background processes
 // Examples of this: ReSync, Compaction
 type BackgroundManagerProcessI interface {
-	Run(options map[string]interface{}, terminator chan struct{}) error
+	Init(options map[string]interface{}, clusterStatus []byte) error
+	Run(options map[string]interface{}, persistClusterStatusCallback updateStatusClusterAwareCallback, terminator chan struct{}) error
 	GetProcessStatus(status BackgroundManagerStatus) ([]byte, error)
 	ResetStatus()
 }
 
+type updateStatusClusterAwareCallback func() error
+
 func (b *BackgroundManager) Start(options map[string]interface{}) error {
 	err := b.markStart()
 	if err != nil {
 		return err
 	}
 
+	var processClusterStatus []byte
+	if b.clusterAwareOptions != nil {
+		processClusterStatus, _, err = b.clusterAwareOptions.bucket.GetRaw(b.clusterAwareOptions.StatusDocID())
+		if err != nil && !base.IsDocNotFoundError(err) {
+			return pkgerrors.Wrap(err, "Failed to get current process status")
+		}
+	}
+
 	b.resetStatus()
+
+	err = b.Process.Init(options, processClusterStatus)
+	if err != nil {
+		return err
+	}
+
+	if b.clusterAwareOptions != nil {
+		go func() {
+			ticker := time.NewTicker(BackgroundManagerStatusUpdateIntervalSecs * time.Second)
+			for {
+				select {
+				case <-ticker.C:
+					err = b.UpdateStatusClusterAware()
+					if err != nil {
+						base.Warnf("Failed to update background manager status: %v", err)
+					}
+
+				case <-b.terminator:
+					return
+				}
+			}
+		}()
+	}
+
 	go func() {
-		defer b.setRunState(BackgroundProcessStateStopped)
-		err := b.Process.Run(options, b.terminator)
+		updateStatusClusterAwareCallback := func() error {
+			if b.clusterAwareOptions != nil {
+				return b.UpdateHeartbeatDocClusterAware()
```

Comment:
Shouldn't this be UpdateStatusClusterAware()?  Assuming yes, can we move the b.clusterAwareOptions check down into that function, and just pass it directly instead of having to wrap with this callback?